warehouse/packaging/models.py

```diff
@@ -344,9 +372,11 @@ def __table_args__(cls):  # noqa
         ForeignKey("projects.id", onupdate="CASCADE", ondelete="CASCADE"),
         nullable=False,
     )
+    project_name = Column(Text)
```

Comment:
We might want to think about ways to avoid having to store the project name on the `Release` model, as this somewhat complicates our models.

This is currently necessary because the project name is otherwise unavailable to the column property to compute the `draft_hash`.

Reply:
It feels like ``draft_hash`` should not be a column property, and it should just be a string that is a derived value. You could have this triggered in the DB or in python.

If we change the mechanism by which the draft hash is computed, we probably don't want to change old hashes, so that their URLs remain stable, so that leads me to think that it should be derived at the point of release creation I think?

Reply:
Or stepping back, what is the value of hashing these values at all? If the inputs are well known it feels like the hash is just there to make the URL more obscure in general, but I don't think they provide much value on their own. Is there any reason the draft release url doesn't just have the project name and version in the url?

If I recall, the original proposal for draft releases used random URLs to prevent people from being able to "guess" the  draft release url, and forcing it to be something that they were given by PyPI. This was an attempt to limit some abuses of draft releases as some sort of release channel or something. But if we're just using urls that anyone can compute knowing nothing but project name + version, I don't see a ton of value in jumping through hopes to embed that rather than just doing something like:

``/draft/$project/$version/``

Reply:
The meain reason that we're calculatin a hash over these values is to avoid accidental usage by regular users. By requiring draft releases be installed with the use of this obfuscated "tokens" which are available and computable by maintainers, it sends the message that you shouldn't try to create them as a user. Avoiding abuse of this feature (as a private release channel for example) could be further mitigated by other means, like deleating the draft reelease after a certain time (think, a day). I do concede that the current implemention has more hoops to jump that I would like.

Reply:
It doesn't feel like there's going to be a huge risk of accidental usage if using it requires typing ``--index-url https://pypi.org/draft/foo/1.0/`` and just happen to know that there exists a draft release of that url? Is there some other mechanism we expect users to enable using a draft release? 

If we really do want to hash it like this, we could probably just do ``/draft/hash($project)/hash($version)/`` yea? Could even do ``/draft/hash($project)hash($version)/`` and just split based on length or use a - or something.

IOW, if we treat the project and version as distinct pieces of data to be hashed rather than trying to combine them and then hash them, we can sidestep all the need to denormalize since it wouldn't be any different than looking up by ``Project.name`` + ``Release.version``.

I'd also suggest not using md5, not because this is security sensitive, but because some Linux distros patch ``hashlib.md5`` to require passing a ``usedforsecurity=False`` attribute to make md5 work at all, and it's simpler to just side step that completely. It should be reasonable enough to use sha256, or if we're worried about the length of the URL and we can mandate 3.6+ (I don't recall what the tools involved support at the moment) you can use blake2 with a custom sized key.

Something like:

```python
import hashlib

hashlib.blake2b(b"adf", digest_size=16).hexdigest()
```

Should produce digests of similar to MD5, you could probably go smaller too?

Could also use an alternative encoding if we're worried about the length of URLs:

```python
import base64
import hashlib

hashlib.blake2b(b"adf", digest_size=16).hexdigest()  # 32 characters
base64.urlsafe_b64encode(hashlib.blake2b(b"adf", digest_size=16).digest()).strip(b"=") .decode("ascii") # 22 characters
```

Anyways, just some food for thought